File: app/src/main/java/antoninovitale/dropcodechallenge/details/BeerDetailsInteractor.java
package antoninovitale.dropcodechallenge.details;

import java.util.List;

import antoninovitale.dropcodechallenge.api.model.Beer;
import antoninovitale.dropcodechallenge.details.section.model.HeaderSectionModel;
import antoninovitale.dropcodechallenge.details.section.model.IngredientSectionModel;
import antoninovitale.dropcodechallenge.details.section.model.MethodSectionModel;
import antoninovitale.dropcodechallenge.details.section.model.Status;

/**
 * Created by antoninovitale on 01/09/2017.
 */
interface BeerDetailsInteractor {
    interface OnStatusCheckListener {

        void onMaltDone(int position);

        void onHopDone(int position);

        void onMethodDone(int position);

        void onMethodStatusChanged(int position, Status status);

    }

    interface OnSectionSetupListener {

        void onHeaderSet(HeaderSectionModel model);

        void onMaltsSet(List<IngredientSectionModel> malts);

        void onHopsSet(List<IngredientSectionModel> hops);

        void onMethodsSet(List<MethodSectionModel> methods);

    }

    void setupBeer(Beer beer, OnSectionSetupListener onSectionSetupListener);

    void checkIngredientStatus(int position, OnStatusCheckListener onStatusCheckListener);

    void checkMethodStatus(int position, OnStatusCheckListener onStatusCheckListener);

    void setMethodDone(int position, OnStatusCheckListener onStatusCheckListener);

    void setMethodTimeElapsed(int position, long millis);

}

File: app/src/main/java/antoninovitale/punkapi/app/details/BeerDetailsInteractor.kt
package antoninovitale.punkapi.app.details

import antoninovitale.punkapi.app.api.model.Beer
import antoninovitale.punkapi.app.details.section.model.HeaderSectionModel
import antoninovitale.punkapi.app.details.section.model.IngredientSectionModel
import antoninovitale.punkapi.app.details.section.model.MethodSectionModel
import antoninovitale.punkapi.app.details.section.model.Status

/**
 * Created by antoninovitale on 01/09/2017.
 */
internal interface BeerDetailsInteractor {
    fun setupBeer(beer: Beer, onSectionSetupListener: OnSectionSetupListener)
    fun checkIngredientStatus(position: Int, onStatusCheckListener: OnStatusCheckListener)
    fun checkMethodStatus(position: Int, onStatusCheckListener: OnStatusCheckListener)
    fun setMethodDone(position: Int, onStatusCheckListener: OnStatusCheckListener)
    fun setMethodTimeElapsed(position: Int, millis: Long)

    interface OnStatusCheckListener {
        fun onMaltDone(positionInSection: Int, globalPosition: Int)
        fun onHopDone(positionInSection: Int, globalPosition: Int)
        fun onMethodDone(positionInSection: Int, globalPosition: Int)
        fun onMethodStatusChanged(positionInSection: Int, globalPosition: Int, status: Status)
    }

    interface OnSectionSetupListener {
        fun onHeaderSet(model: HeaderSectionModel)
        fun onMaltsSet(malts: List<IngredientSectionModel>)
        fun onHopsSet(hops: List<IngredientSectionModel>)
        fun onMethodsSet(methods: List<MethodSectionModel>)
    }
}